tests/sample/reflect_packet.c

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "bpf_helpers.h"
+#include "ebpf.h"
+#include "xdp_tests_common.h"
+
+void
+copy_mac_address(mac_address_t destination, mac_address_t source)
+{
+    for (int i = 0; i < sizeof(mac_address_t); i++)
+        destination[i] = source[i];
+}
+
+void
+swap_mac_addresses(ETHERNET_HEADER* eth_hdr)
+{
+    mac_address_t temp = {0};
+    copy_mac_address(temp, eth_hdr->Destination);
+    copy_mac_address(eth_hdr->Destination, eth_hdr->Source);
+    copy_mac_address(eth_hdr->Source, temp);
+}
+
+void
+swap_ipv4_addresses(IPV4_HEADER* ipv4_hdr)
+{
+    uint32_t temp = ipv4_hdr->DestinationAddress;
+    ipv4_hdr->DestinationAddress = ipv4_hdr->SourceAddress;
+    ipv4_hdr->SourceAddress = temp;
+}
+
+void
+copy_ipv6_address(ipv6_address_t destination, ipv6_address_t source)
+{
+    for (int i = 0; i < sizeof(ipv6_address_t); i++)
+        destination[i] = source[i];
+}
+
+void
+swap_ipv6_addresses(IPV6_HEADER* ipv6_hdr)
+{
+    ipv6_address_t temp = {0};
+    copy_ipv6_address(temp, ipv6_hdr->DestinationAddress);
+    copy_ipv6_address(ipv6_hdr->DestinationAddress, ipv6_hdr->SourceAddress);
+    copy_ipv6_address(ipv6_hdr->SourceAddress, temp);
+}
+
+//
+// This eBPF program intercepts inbound UDP packets destined to port REFLECTION_TEST_PORT and "reflects" it back
+// by swapping the MAC and IP addresses.
+//
+SEC("xdp")
+int
+reflect_packet(xdp_md_t* ctx)
+{
+    int rc = XDP_PASS;
+
+    ETHERNET_HEADER* eth_hdr = NULL;
+    char* next_header = (char*)ctx->data;
+    if ((char*)next_header + sizeof(ETHERNET_HEADER) > (char*)ctx->data_end)
+        goto Done;
+    eth_hdr = (ETHERNET_HEADER*)next_header;
+    next_header = (char*)(eth_hdr + 1);
+    if (eth_hdr->Type == ntohs(ETHERNET_TYPE_IPV4)) {
+        if ((char*)next_header + sizeof(IPV4_HEADER) > (char*)ctx->data_end)
+            goto Done;
+        // IPv4.
+        IPV4_HEADER* ipv4_hdr = (IPV4_HEADER*)next_header;
+        next_header = (char*)(ipv4_hdr + 1);
```

Comment:
Since this is a sample, the code should not have bugs.  The bug here is that if IHL != 5, this code is wrong.  Instead here should be something like:
```suggestion
        next_header = ((char*)ipv4_header) + sizeof(uint32_t) * ipv4_header->HeaderLength;
```

Reply:
I didn't notice a change here to fix this bug yet.

Reply:
For CI/CD unit tests, the ipv4 header length is not set. This is a sample eBPF program and not something in production. This is just for testing XDP_TX works and packets are reflected back in the scoped scenario of testing.